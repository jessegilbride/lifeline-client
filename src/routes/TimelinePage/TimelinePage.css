.timeline-container {
  --timeline-width: 4px;
}

.timeline-container {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  position: relative;
  max-width: 800px;
  min-height: calc(80vh - var(--top-spacing));
}

.timeline-container::after {
  position: absolute;
  left: calc(50% - var(--timeline-width)/2 );
  background-color: red;
  height: 95%;
  width: var(--timeline-width);
  content: '';
}

.btn-fixed {
  position: fixed;
  /* top: calc(var(--top-spacing) + 40px); */
  bottom: 60px;
  right: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  height: 80px;
  width: 80px;
  border-radius: 50%;
  text-decoration: none;
  font-family: monospace;
  padding: 0.5rem;
  z-index: 1;
}

.add-item {
  background-color: var(--blue);
  color: white;
  font-size: 1rem;
}

@media (min-width: 48em) {
  .timeline-container::after {
    position: absolute;
    /* left: calc(100% - var(--timeline-width)/2); */
    left: calc(100%);
    height: 95%;
    width: var(--timeline-width);
    content: '';
  }

  .btn-fixed {
    bottom: 100px;
    right: 80px;
    height: 100px;
    width: 100px;
  }
}