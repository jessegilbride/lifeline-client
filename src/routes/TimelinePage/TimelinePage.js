import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import TimelineApiService from '../../services/timeline-api-service';
import TimelineContext from '../../contexts/TimelineContext';
import TimelineItem from '../../components/TimelineItem/TimelineItem';
import './TimelinePage.css';

export default class TimelinePage extends Component {

  static contextType = TimelineContext;

  componentDidMount() {
    TimelineApiService.getTimelines()
      .then((timeline) => {
        // console.log('getTimelines() ... ', timeline)

        if(timeline.length > 0) {
          this.context.setTimeline(timeline)
          return TimelineApiService.getItems(timeline[0].id)
        } 
        else {this.props.history.push('/create-timeline')}

        return [];

      })
      .then((items) => this.context.setItems(items))
      .catch(error => {
        console.log(error)
        // return this.context.setError // use local state
      });

    // const line_id = this.context.line.id;
    // const line_id = 1; // hard-coded, needs to be dynamic

    // TimelineApiService.getItems(line_id)
      // .then((items) => this.context.setItems(items))
      // .catch(error => {
        // console.log(error)
        // return this.context.setError // use local state
      // })
  }

  sortTheItems(items) {
    const itemsCopy = [...items];
    
    return itemsCopy.sort((a, b) => {
      // console.log('a: ', a)
      // console.log('a.entry_date: ', a)
      // console.log('b.entry_date: ', b)
      // console.log('b.entry_date: ', b)
      return Date.parse(a.entry_date) - Date.parse(b.entry_date)
    })
  }

  renderTimelineItems() {
    const { items = [] } = this.context;
    
    const sortedItems = this.sortTheItems(items);
    // console.log('sortedItems... ', sortedItems)
    // console.log('items... ', items)

    return sortedItems.map((item, idx) => <TimelineItem itemData={item} key={idx} />).reverse()

    // {id, line_id, title, content, entry_date}
    // return items.map((item, idx) => <TimelineItem itemData={item} key={idx} />)
  }

  render() {
    const { timeline, error } = this.context;
    return (
      <Fragment>
        <section className="page-width-container page-top-banner">
          <h2 className='page-heading'>{timeline && timeline.line_name}</h2>
        </section>
        <section className="page-width-container timeline-container">
          <Link to='/create-timeline-item' className='btn btn-fixed add-item' aria-label='add timeline entry'>add item</Link>
          {error ? (
            <p className='red'>There was an error, try again</p>
          ) : (
            this.renderTimelineItems()
          )}
          {/* {this.renderTimelineItems()} */}
        </section>
      </Fragment>
    );
  }
}
