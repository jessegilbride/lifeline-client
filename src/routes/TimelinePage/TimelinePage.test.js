import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from "react-router-dom";
import renderer from 'react-test-renderer';
import TimelinePage from './TimelinePage';

describe('TimelinePage component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
    <MemoryRouter><TimelinePage /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(
      <MemoryRouter><TimelinePage name="TimelinePage" /></MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

});
