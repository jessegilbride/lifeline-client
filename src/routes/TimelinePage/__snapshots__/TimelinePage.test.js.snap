// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelinePage component renders the UI as expected 1`] = `
Array [
  <section
    className="page-width-container page-top-banner"
  >
    <h2
      className="page-heading"
    />
  </section>,
  <section
    className="page-width-container timeline-container"
  >
    <a
      aria-label="add timeline entry"
      className="btn btn-fixed add-item"
      href="/create-timeline-item"
      onClick={[Function]}
    >
      add item
    </a>
  </section>,
]
`;
